File: lingting-elasticsearch/src/main/java/live/lingting/framework/elasticsearch/ElasticsearchProvider.java
package live.lingting.framework.elasticsearch;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.json.JsonpMapper;
import co.elastic.clients.json.jackson.JacksonJsonpMapper;
import co.elastic.clients.transport.ElasticsearchTransport;
import co.elastic.clients.transport.rest_client.RestClientTransport;
import com.fasterxml.jackson.databind.ObjectMapper;
import live.lingting.framework.elasticsearch.datascope.ElasticsearchDataPermissionHandler;
import live.lingting.framework.jackson.JacksonUtils;
import lombok.experimental.UtilityClass;
import org.apache.http.HttpHost;
import org.elasticsearch.client.RestClient;

import java.util.Arrays;
import java.util.function.Function;

/**
 * @author lingting 2024-03-06 20:00
 */
@UtilityClass
public class ElasticsearchProvider {

	public static RestClient restClient(String... hosts) {
		return restClient(Arrays.stream(hosts).map(HttpHost::create).toArray(HttpHost[]::new));
	}

	public static RestClient restClient(HttpHost... hosts) {
		return RestClient.builder(hosts).build();
	}

	public static JsonpMapper jacksonMapper() {
		return jacksonMapper(JacksonUtils.getMapper());
	}

	public static JsonpMapper jacksonMapper(ObjectMapper mapper) {
		return new JacksonJsonpMapper(mapper);
	}

	public static RestClientTransport transport(RestClient restClient) {
		return transport(restClient, jacksonMapper());
	}

	public static RestClientTransport transport(RestClient restClient, JsonpMapper jsonpMapper) {
		return new RestClientTransport(restClient, jsonpMapper);
	}

	public static ElasticsearchClient client(HttpHost... hosts) {
		return client(restClient(hosts));
	}

	public static ElasticsearchClient client(RestClient restClient) {
		return client(transport(restClient));
	}

	public static ElasticsearchClient client(ElasticsearchTransport transport) {
		return new ElasticsearchClient(transport);
	}

	public static <T> ElasticsearchApi<T> api(Class<T> cls, Function<T, String> idFunc,
			ElasticsearchProperties properties, ElasticsearchDataPermissionHandler handler,
			ElasticsearchClient client) {
		return new ElasticsearchApi<>(cls, idFunc, properties, handler, client);
	}

	public static <T> ElasticsearchApi<T> api(String index, Class<T> cls, Function<T, String> idFunc,
			ElasticsearchProperties properties, ElasticsearchDataPermissionHandler handler,
			ElasticsearchClient client) {
		return new ElasticsearchApi<>(index, cls, idFunc, properties, handler, client);
	}

}

File: lingting-elasticsearch/src/main/kotlin/live/lingting/framework/elasticsearch/ElasticsearchProvider.kt
package live.lingting.framework.elasticsearch

import co.elastic.clients.elasticsearch.ElasticsearchClient
import co.elastic.clients.json.JsonpMapper
import co.elastic.clients.json.jackson.JacksonJsonpMapper
import co.elastic.clients.transport.ElasticsearchTransport
import co.elastic.clients.transport.rest_client.RestClientTransport
import com.fasterxml.jackson.databind.ObjectMapper
import live.lingting.framework.jackson.JacksonUtils
import org.apache.http.HttpHost
import org.elasticsearch.client.RestClient

/**
 * @author lingting 2024-03-06 20:00
 */
object ElasticsearchProvider {

    @JvmStatic
    fun restClient(vararg hosts: String): RestClient {
        val map = hosts.mapNotNull() { HttpHost.create(it) }
        return restClient(*map.toTypedArray())
    }

    @JvmStatic
    fun restClient(vararg hosts: HttpHost): RestClient {
        return RestClient.builder(*hosts).build()
    }

    @JvmStatic
    @JvmOverloads
    fun jacksonMapper(mapper: ObjectMapper = JacksonUtils.mapper): JsonpMapper {
        return JacksonJsonpMapper(mapper)
    }

    @JvmStatic
    @JvmOverloads
    fun transport(restClient: RestClient, jsonpMapper: JsonpMapper = jacksonMapper()): RestClientTransport {
        return RestClientTransport(restClient, jsonpMapper)
    }

    @JvmStatic
    fun client(vararg hosts: HttpHost): ElasticsearchClient {
        return client(restClient(*hosts))
    }

    @JvmStatic
    fun client(restClient: RestClient): ElasticsearchClient {
        return client(transport(restClient))
    }

    @JvmStatic
    fun client(transport: ElasticsearchTransport): ElasticsearchClient {
        return ElasticsearchClient(transport)
    }

}
